### ISSUE TITLE
Tinyraytracer in Python

### ISSUE THREAD

**Issue by masscry** [ssloy/tinyraytracer#16]
Hello!

Thank you for very interesting materials!

I ported most of code to Python (except bunny rendering...).

[python-tinyraytracer](https://github.com/masscry/python-tinyraytracer)
It is a bit dirty and not very pythonic, but it works and produces same results.

![out](https://user-images.githubusercontent.com/21346557/52068541-eab72b80-258d-11e9-96e7-5c6ec73468ac.png)

**Comment by ssloy** [comment]
Good job, thank you for sharing! One minor comment:

<a href="https://github.com/masscry/python-tinyraytracer/commit/e698d1e05d25d022730a6247d0c7e59bcf7124cd#diff-b2276c8bfd2a9da6c726ba2f68e4058cR148">in this place</a>  i'd put spherical coordinates instead:

```python
x = int((math.atan2(norm_dir.z, norm_dir, x)/(2*math.pi) + 0.5)*env_width)
y = int(math.acos(norm_dir.y)/math.pi*env_height)

```

**Comment by masscry** [comment]
Nice! That is much [better](https://github.com/masscry/python-tinyraytracer/commit/e35549736fc6010ffd19270b4c20a19371e54758)!

![out](https://user-images.githubusercontent.com/21346557/52111422-b89cdc80-2614-11e9-81d9-3622f58e8d2f.png)

I used environment map in low resolution, so background becomes too blocky without filtering.

**Comment by masscry** [comment]
Rendering time skyrocketed (~30 minutes) with duck mesh on scene, but with bounding box I got reasonable ~2 minutes per draw.

![out](https://user-images.githubusercontent.com/21346557/52127271-b0f42c80-2642-11e9-8aa8-2f47d5cfafae.png)

**Comment by ssloy** [comment]
Yup, exponential complexity. I give this in my lectures to motivate for <a href="https://github.com/ssloy/tinyrenderer">the rasterization technique</a>.

I guess that Python makes it a little bit more expensive, C++ implementation takes a couple of minutes without the bounding box.
